File: benchmarks/hashtable/bench_probing_hashtable.cpp
/*******************************************************************************
 * benchmarks/hashtable/bench_probing_hashtable.cpp
 *
 * Part of Project Thrill - http://project-thrill.org
 *
 * Copyright (C) 2015 Alexander Noe <aleexnoe@gmail.com>
 * Copyright (C) 2015 Matthias Stumpp <mstumpp@gmail.com>
 *
 * All rights reserved. Published under the BSD-2 license in the LICENSE file.
 ******************************************************************************/

#include <thrill/common/cmdline_parser.hpp>
#include <thrill/common/stats_timer.hpp>
#include <thrill/core/reduce_post_stage.hpp>
#include <thrill/core/reduce_pre_stage.hpp>
#include <thrill/data/block_writer.hpp>
#include <thrill/data/discard_sink.hpp>
#include <thrill/data/file.hpp>

#include <algorithm>
#include <climits>
#include <cmath>
#include <functional>
#include <iostream>
#include <iterator>
#include <limits>
#include <numeric>
#include <random>
#include <string>
#include <utility>
#include <vector>

using IntPair = std::pair<int, int>;

using namespace thrill; // NOLINT

int main(int argc, char* argv[]) {

    common::CmdlineParser clp;

    clp.SetVerboseProcess(false);

    std::string title = "";
    clp.AddString('t', "title", "T", title,
                  "Load in byte to be inserted");

    unsigned int size = 5000000;
    clp.AddUInt('s', "size", "S", size,
                "Load in byte to be inserted");

    unsigned int workers = 100;
    clp.AddUInt('w', "workers", "W", workers,
                "Open hashtable with W workers, default = 1.");

    double max_partition_fill_rate = 0.5;
    clp.AddDouble('f', "max_partition_fill_rate", "F", max_partition_fill_rate,
                  "Open hashtable with max_partition_fill_rate, default = 0.5.");

    double table_rate = 1.0;
    clp.AddDouble('r', "table_rate", "R", table_rate,
                  "Open hashtable with max_partition_fill_rate, default = 1.0.");

    unsigned int byte_size = 5000000;
    clp.AddUInt('m', "max_num_items_table", "M", byte_size,
                "Table size, default = 500000000.");

    static const bool full_reduce = false;

    if (!clp.Process(argc, argv)) {
        return -1;
    }

    ///////
    // strings mode
    ///////

    api::Run([&](api::Context& ctx) {

                 auto key_ex = [](size_t in) { return in; };

                 auto red_fn = [](size_t in1, size_t in2) {
                                   (void)in2;
                                   return in1;
                               };

                 size_t num_items = size / sizeof(std::pair<size_t, size_t>);

                 std::default_random_engine rng(std::random_device { } ());
                 std::uniform_int_distribution<size_t> dist(1, std::numeric_limits<size_t>::max());

                 data::BlockPool block_pool(workers);
                 std::vector<data::File> sinks;
                 std::vector<data::File::DynWriter> writers;
                 for (size_t w = 0; w < workers; ++w) {
                     sinks.emplace_back(block_pool, w);
                     writers.emplace_back(sinks[w].GetDynWriter());
                 }

                 core::ReducePreProbingStage<size_t, size_t, size_t, decltype(key_ex), decltype(red_fn), true,
                                             core::ReduceByHashKey<size_t>,
                                             std::equal_to<size_t> >
                 table(ctx, workers, key_ex, red_fn, writers, core::ReduceByHashKey<size_t>(),
                       0, 0, byte_size);

                 common::StatsTimer<true> timer(true);

                 for (size_t i = 0; i < num_items; i++)
                 {
                     table.Insert(dist(rng));
                 }

                 table.Flush();

                 timer.Stop();

                 std::cout << "RESULT" << " benchmark=" << title << " size=" << size << " byte_size=" << byte_size << " workers="
                           << workers << " max_partition_fill_rate=" << max_partition_fill_rate
                           << " table_rate_multiplier=" << table_rate << " full_reduce=" << full_reduce << " final_reduce=true"
                           << " time=" << timer.Milliseconds() << std::endl;
             });

    return 0;
}

/******************************************************************************/

File: benchmarks/hashtable/bench_probing_hashtable.py
#!/usr/bin/env python
##########################################################################
# benchmarks/hashtable/bench_probing_hashtable.py
#
# Part of Project Thrill - http://project-thrill.org
#
#
# All rights reserved. Published under the BSD-2 license in the LICENSE file.
##########################################################################

import os
import sys
import subprocess
import random
import numpy

result_dir = "./bench_probing_hashtable"
if not os.path.exists(result_dir): os.makedirs(result_dir)

def bench(workers, sizes, intervals, max_partition_fill_rates, num_items_per_partitions, table_sizes, num_runs):
    for worker in workers:
        for size in sizes:
            for interval in intervals:
                for max_partition_fill_rate in max_partition_fill_rates:
                    for num_items_per_partition in num_items_per_partitions:
                        for table_size in table_sizes:
                            with open(result_dir + "/" + "4" + "_" + str(worker) + "_" + str(size) + "_" + str(interval[0]) + "-" + str(interval[1]) + "_" + str(max_partition_fill_rate) + "_" + str(num_items_per_partition) + "_" + str(table_size), "w+") as file1:
                                times = []
                                flushes = []
                                items = []
                                for _ in range(num_runs):
                                    process = subprocess.Popen(['../../build/benchmarks/hashtable_bench_probing_hashtable', '-s', str(size), '-w', str(worker), '-l', str(interval[0]), '-u', str(interval[1]), '-f', str(max_partition_fill_rate), '-i', str(num_items_per_partition), '-t', str(table_size)], stdout=subprocess.PIPE)
                                    process.wait()
                                    out = process.communicate()[0]
                                    out_s = out.split();
                                    times.append(float(out_s[0]))
                                    flushes.append(float(out_s[1]))
                                    items.append(float(out_s[2]))
                                time = numpy.median(times)
                                flush = numpy.median(flushes)
                                item = numpy.median(items)
                                print str(worker) + "_" + str(interval[0]) + "-" + str(interval[1]) + "_" + str(max_partition_fill_rate) + "_" + str(num_items_per_partition) + "_" + str(table_size) + ": " + str(time) + " " + str(flush) + " " + str(item) + " " + str(item * (pow(10,6)/time)) + " " + str(flush * (pow(10,6)/time))
                                file1.write(str(interval[0]) + "-" + str(interval[1]) + " " + str(time) + " " + str(flush) + " " + str(item) + " " + str(item * (pow(10,6)/time)) + " " + str(flush * (pow(10,6)/time)) + "\n")
                            file1.close()

num_runs = 5
bench([100], [2000000000], [[5, 15]], [0.5], [1], [500000000], num_runs);
#5 not applied

##########################################################################
